### DISCUSSION TITLE
Flashing in DFU mode for UART

### DISCUSSION THREAD

**Discussion by arunbhaarathi** [bitcraze/discussions#80] [Q&A] [answered]
Q1) I am trying to control the crazyflie with a raspberrypi zero using UART. The document provided [here](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/development/uart_communication/) says  "the firmware has to be compiled with the UART2_LINK=1 flag (e.g. make UART2_LINK=1) and flashed to the Crazyflie, to enable UART". I am unable to figure out where to change the UART2_LINK parameter to enable UART.
 
Q2) I don't have access to the crazyradio usb dongle hence tried updating the firmware over USB in DFU mode using the instructions provided [here](https://github.com/bitcraze/crazyflie-firmware/blob/master/docs/development/dfu.md). How to flash the updated firmware to the CF in DFU mode? 

TIA

**Comment by tobbeanton** [comment, marked as answer]
Q1) The build system has changed a bit since the guide was written and is now using kbuild and the guide way doesn't work anymore. It should be added as a kbuild choice instead. I added [#1041](https://github.com/bitcraze/crazyflie-firmware/issues/1041) for this. Until then you can add
`#define UART2_LINK_COMM` to a line in config.h and recompile the FW. I hope it is still working!

**Reply by arunbhaarathi** [reply]
Thank you @tobbeanton. I tried your solution, it didn't work. I configured the UART ports as mentioned in the guide, "ttyAMA0" is set to serial 0. I ran the test code mentioned in the [guide](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/development/uart_communication/), I don't get any output or errors, the screen is frozen. I tried adding a print statement inside each loop to check the functioning of the code. The main function is called but it doesn't enter into the following loop,
```
 with SyncCrazyflie(URI) as scf:
        # We take off when the commander is created
        with MotionCommander(scf) as mc:
            print('Taking off!')
            time.sleep(0.1)
            # We land when the MotionCommander goes out of scope
            print('Landing!')

```
It would be great if you can help me with it ASAP.

**Reply by tobbeanton** [reply]
I tried this using a FTDI cable connected to the computer and TX2, RX2 on the Crazyflie and connecting using the cfclient. I had to change:
```
    # Initialize the low-level drivers including the serial driver
    cflib.crtp.init_drivers(enable_serial_driver=True)
```
in the cfclient main.py to enable the tty* in the connect drop-down menu. Too me the startup is unstable and sometimes hangs ,but if it starts, then it seems to work. 
The Kbuild option is now merged and if you run the latest on the master branch you should be able to enable the communication->CRTP_OVER_UART2 menuconfig option. I have no possibility to try the raspberry way at the moment.

**Reply by arunbhaarathi** [reply]
@tobbeanton It's working now, thank you.

**Comment by tobbeanton** [comment]
Q2) Did the guide not work? You have to elaborate.

**Reply by arunbhaarathi** [reply]
This is one is fixed and working, thanks.
